CLOUDP-212065: fix: add 2.0.0 major version of AKO

s-urbaniak opened the pull request:
<!--
Thanks for contributing to MongoDB CLI!

Before you submit your pull request, please review our contribution guidelines:
https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md

Please fill out the information below to help speed up the review process
and getting you pull request merged!
-->

## Proposed changes

Starting from the changes present in this PR, rendering AKO v1.x resources is not supported any more. Starting from here, rendering AKO v2.x resources is possible only.

_Jira ticket:_ CLOUDP-210683

<!--
What MongoDB CLI issue does this PR address? (for example, #1234), remove this section if none.
-->

## Checklist

<!--
Check the boxes that apply. If you're unsure about any of them, don't hesitate to ask!
We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [ ] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [ ] I have run `make fmt` and formatted my code

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.

Alternatively, if this is a very minor, and self-explanatory change, feel free to remove this section.
-->

s-urbaniak commented:
@mongodb/docs-cloud-team please be aware of the changes mentioned in the PR title. Please let me know if you have further questions, thank you!

gssbzn commented on test/e2e/atlas/helper_test.go:
```diff
@@ -270,6 +270,7 @@ func deployClusterForProject(projectID, tier string, enableBackup bool) (string,
 
 func e2eTier() string {
 	tier := e2eClusterTier
+	// tier := e2eSharedClusterTier
```
```suggestion
```

gssbzn commented on go.mod:
```diff
@@ -58,14 +58,15 @@ require (
 	k8s.io/api v0.26.4
 	k8s.io/apiextensions-apiserver v0.26.1
 	k8s.io/apimachinery v0.26.4
+	k8s.io/apiserver v0.26.1
```
this lib is missing an owner

s-urbaniak replied:
fixed

gssbzn commented on test/e2e/atlas/operator_helper_test.go:
```diff
@@ -167,7 +158,7 @@ func (oh *operatorHelper) stopOperator() {
 	}
 
-	deployment.Spec.Replicas = toptr.MakePtr(int32(0))
+	deployment.Spec.Replicas = akov2toptr.MakePtr(int32(0))
```
we have our own that I'd like to use instead of more dependencies with kube
```suggestion
	deployment.Spec.Replicas = pointer.Get(0)
```

gssbzn commented on test/e2e/atlas/kubernetes_config_apply_test.go:
```diff
@@ -67,10 +69,22 @@ func TestKubernetesConfigApply(t *testing.T) {
 			"kubernetes",
 			"config",
 			"apply",
+			"--targetNamespace", "e2e-autodetect-parameters",
```
is this flag now mandatory or was it added here for the test

s-urbaniak replied:
it was added as concurrently executing tasks stomped on each other in the same namespace.

s-urbaniak commented on test/e2e/atlas/kubernetes_config_generate_test.go:
```diff
@@ -1016,91 +1067,114 @@ func referenceServerless(name, region, namespace, projectName string, labels map
+					Namespace: project.Namespace,
+				},
+				VictorOpsSecretRef: akov2common.ResourceRefNamespaced{
+					Name:      resources.NormalizeAtlasName(fmt.Sprintf("%s-victor-ops-credentials-%d", project.Name, i), dictionary),
```
note to myself: check if this is still executed, this should use generated names.

igor-karpukhin commented on internal/kubernetes/operator/secrets/secrets.go:
```diff
@@ -21,27 +21,67 @@ import (
-				TypeLabelKey: CredLabelVal,
+type AtlasSecretBuilder func() (*corev1.Secret, map[string]string)
+
+func NewAtlasSecretBuilder(name, namespace string, dictionary map[string]string) AtlasSecretBuilder {
```
Interesting approach :) This can also be implemented with a struct, eliminating the `Build()` method at the end.

s-urbaniak replied:
Indeed, it was something real quick, I don't think we need anything fancy here. If we're going to need more initialising variables we can also introduce a struct. The `Build` method is just to denote that this is a builder so its implementation can vary independently.

igor-karpukhin commented on test/e2e/atlas/operator_helper_test.go:
```diff
@@ -167,7 +158,7 @@ func (oh *operatorHelper) stopOperator() {
 	}
 
-	deployment.Spec.Replicas = toptr.MakePtr(int32(0))
+	deployment.Spec.Replicas = pointer.Get(0)
```
This has to be `int32` not `int`

s-urbaniak replied:
@gssbzn fyi ^ probably that's why the other method was used.

gssbzn replied:
pointer.Get is a generic
```
pointer.Get[int32](0)
```

s-urbaniak replied:
fixed

igor-karpukhin reviewed (commented):
There is one issue with the type check. LGTM in general, if the CI is green.

gssbzn commented on internal/kubernetes/operator/features/crds.go:
```diff
@@ -47,24 +47,7 @@ var (
-			ResourceAtlasDataFederation,
-		},
-		"1.9.0": {
+		"2.0.0": {
```
[q] we are dropping support for v1 then? I'm ok just confirming

gssbzn commented on internal/kubernetes/operator/features/crds_test.go:
```diff
@@ -273,9 +273,9 @@ func Test_CRDCompatibleVersion(t *testing.T) {
+		latestOperatorSemver, err := semver.NewVersion(LatestOperatorMajorVersion)
 		require.NoError(t, err)
-		operatorVersion := semver.New(crdVersion.Major(), crdVersion.Minor()-1, 2, "", "")
+		operatorVersion := semver.New(latestOperatorSemver.Major(), latestOperatorSemver.Minor(), 2, "", "")
```
the original -1 was on the patch, which currently was 2.0.2, so it would produce 2.-1.2, my -1 is on major, producing 1.0.2 ;)

s-urbaniak replied:
fixed

s-urbaniak commented:
@gssbzn i hopefully addressed all your concerns, thanks for the feedback, PTAL. Having said that I prefixed this PR as WIP (work in progress). We unfortunately found another bug yesterday in AKO v2.0 which we are fixing. The implied changes here most likely will be:
- We will not advertise `2.0.0` in the Atlas-CLI, just the `2.0.1` release including the bug fixes
- reference the `2.0.1` images

cc @igor-karpukhin

gssbzn reviewed (dismissed):
Changes LGTM you can decide when is best to merge this once the operator fixes are out but I'm not blocking the PR

igor-karpukhin reviewed (dismissed):
Great work! 👍

s-urbaniak commented:
before we merge, I'd like to implement one last optimization. Currently, we always generate empty secret templates. That's not really necessary unless a concrete alert integration is configured in Atlas.

We can consider the integration type name to see if we actually need to create a secret template.

igor-karpukhin reviewed (dismissed):
Looks good 👍
